ostree/ostree_src.go

```diff
@@ -0,0 +1,188 @@
+	b := fmt.Sprintf("ociimage/%s", blob)
+	// Use golang bindings once they support to read metadata
+	out, err := exec.Command("ostree", "show", "--repo", s.ref.repo, "--print-metadata-key=tarsplit.output", b).CombinedOutput()
+	if err != nil {
+		if strings.Index(string(out), "No such metadata key") >= 0 {
+			return nil, nil
+		}
+		return nil, err
+	}
+	data := parseGVariantString(string(out))
+	return base64.StdEncoding.DecodeString(data)
+}
+
+// GetManifest returns the image's manifest along with its MIME type (which may be empty when it can't be determined but the manifest is available).
+// It may use a remote (= slow) service.
+func (s *ostreeImageSource) GetManifest() ([]byte, string, error) {
+	b := fmt.Sprintf("ociimage/%s", s.ref.branchName)
+	// Use golang bindings once they support to read metadata
+	out, err := exec.Command("ostree", "show", "--repo", s.ref.repo, "--print-metadata-key=docker.manifest", b).Output()
+	if err != nil {
+		return nil, "", err
+	}
+	m := []byte(parseGVariantString(string(out)))
+	return m, manifest.GuessMIMEType(m), nil
+}
+
+func (s *ostreeImageSource) GetTargetManifest(digest digest.Digest) ([]byte, string, error) {
+	return nil, imgspecv1.MediaTypeImageManifest, nil
+}
+
+func (s *ostreeImageSource) checkout(dest string, checksum string) error {
+	opts := otbuiltin.NewCheckoutOptions()
+	opts.UserMode = true
+	commit := fmt.Sprintf("ociimage/%s", checksum)
+	return otbuiltin.Checkout(s.ref.repo, dest, commit, opts)
+}
+
+// GetBlob returns a stream for the specified blob, and the blob's size.
+func (s *ostreeImageSource) GetBlob(info types.BlobInfo) (io.ReadCloser, int64, error) {
+	blob := info.Digest.Hex()
+
+	dir, err := ioutil.TempDir(s.tmpDir, string(info.Digest))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	layer := fmt.Sprintf("%s/layer", dir)
+
+	err = s.checkout(layer, blob)
+	if err != nil {
+		os.RemoveAll(dir)
+		return nil, 0, err
+	}
+
+	layerSize, err := s.getLayerSize(blob)
+	if err != nil {
+		os.RemoveAll(dir)
+		return nil, 0, err
+	}
+
+	tarsplit, err := s.getTarSplitData(blob)
+	if err != nil {
+		os.RemoveAll(dir)
+		return nil, 0, err
+	}
+
+	// if tarsplit is nil we are looking at the manifest.  Return directly the file in layer/content
+	if tarsplit == nil {
+		stream, err := os.OpenFile(fmt.Sprintf("%s/content", layer), os.O_RDONLY, 0)
+		if err != nil {
+			os.RemoveAll(dir)
+			return nil, 0, err
+		}
+		return ioutils.NewReadCloserWrapper(stream, func() error {
+			return os.RemoveAll(dir)
+		}), layerSize, nil
+	}
+
+	mf := bytes.NewReader(tarsplit)
+	mfz, err := gzip.NewReader(mf)
+	if err != nil {
+		os.RemoveAll(dir)
+		return nil, 0, err
+	}
+	defer mfz.Close()
+	metaUnpacker := storage.NewJSONUnpacker(mfz)
+
+	ots := asm.NewOutputTarStream(storage.NewPathFileGetter(layer), metaUnpacker)
+
+	pipeReader, pipeWriter := io.Pipe()
+	gzipWriter := gzip.NewWriter(pipeWriter)
+	go func() {
+		io.Copy(gzipWriter, ots)
+		gzipWriter.Close()
+		pipeWriter.Close()
```

Comment:
Hum, trying to re-create a compressed blob by re-creating the uncompressed tarstream and compressing it doesn’t work long-term and is not supportable.  (We did try to do this earlier, for `containers-storage`, IIRC; it failed after a few months.)  The `gzip` bitstream is not stable over Golang versions, meaning that this will not be able to reproduce the compressed blobs of images compressed by earlier Golang versions, and after any future Golang upgrade a previously working `GetBlob` can start returning different data; in either case the blob won’t match the digest recorded in the manifest.

For me, this inability to handle compressed data is **by no means** a blocker for this PR; `containers-storage` has the same problem, and that will need to be solved by editing the manifest.  See #305 and the conversations in there.

What to do for now, though? I would be fine with this PR just skipping the compression step, meaning that the `ostree:` source produces usable images only for originally uncompressed images, and dealing with the compression as a follow-up.

That follow-up can probably be built around the tentative c/i/manifest subpackage being explored in #305.

Cc: @nalind

Reply:
it is something that is biting us also when we import layers from the local Docker engine, as we compute the checksum using whatever gzip implementation is used by the golang version used to build https://github.com/projectatomic/atomic/blob/master/dockertar-sha256-helper.go.  In anycase, we should probably use `skopeo copy` in atomic for accessing the local Docker engine.